File: api-net/Controllers/SmsController.cs
using LiftOtonom.Api.Data;
using LiftOtonom.Api.Models;
using LiftOtonom.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace LiftOtonom.Api.Controllers;

[ApiController]
[Authorize]
[Route("api/v1/sms")]
public class SmsController(AppDbContext db, ISmsSender sms) : ControllerBase
{
    public record SendDto(string Phone, string Message);
    public record PrefDto(bool? MaintenanceReminder, bool? MaintenanceCompleted, bool? NewFault, bool? FaultResolved,
        bool? InvoiceCreated, bool? PaymentReceived, bool? TseExpiry, int? ReminderDaysBefore);

    private long TenantId => db.CurrentTenantId!.Value;

    [HttpGet("balance")]
    public async Task<IActionResult> Balance()
    {
        var tenant = await db.Tenants.FindAsync(TenantId);
        return Ok(new { balance = tenant!.SmsBalance, sent_total = await db.SmsLogs.CountAsync() });
    }

    [HttpGet("history")]
    public async Task<IActionResult> History([FromQuery(Name = "per_page")] int perPage = 25, [FromQuery] int page = 1) =>
        Ok(await db.SmsLogs.OrderByDescending(s => s.Id).ToPagedAsync(page, perPage));

    [HttpPost("send")]
    public async Task<IActionResult> Send(SendDto dto)
    {
        var tenant = await db.Tenants.FindAsync(TenantId);
        if (tenant!.SmsBalance <= 0) throw new ApiException(402, "SMS bakiyeniz yetersiz.");

        var phone = PhoneHelper.Normalize(dto.Phone);
        await sms.SendAsync(phone, dto.Message, "manual");
        tenant.SmsBalance -= 1;
        await db.SaveChangesAsync();

        return Ok(new { message = "SMS gönderildi.", remaining = tenant.SmsBalance });
    }

    [HttpGet("preferences")]
    public async Task<IActionResult> GetPreferences() => Ok(await Prefs());

    [HttpPut("preferences")]
    public async Task<IActionResult> UpdatePreferences(PrefDto dto)
    {
        var p = await Prefs();
        if (dto.MaintenanceReminder is { } a) p.MaintenanceReminder = a;
        if (dto.MaintenanceCompleted is { } b) p.MaintenanceCompleted = b;
        if (dto.NewFault is { } c) p.NewFault = c;
        if (dto.FaultResolved is { } d) p.FaultResolved = d;
        if (dto.InvoiceCreated is { } e) p.InvoiceCreated = e;
        if (dto.PaymentReceived is { } f) p.PaymentReceived = f;
        if (dto.TseExpiry is { } g) p.TseExpiry = g;
        if (dto.ReminderDaysBefore is { } h) p.ReminderDaysBefore = h;
        p.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync();
        return Ok(p);
    }

    private async Task<SmsPreference> Prefs()
    {
        var p = await db.SmsPreferences.FirstOrDefaultAsync(x => x.TenantId == TenantId);
        if (p == null)
        {
            p = new SmsPreference { TenantId = TenantId, CreatedAt = DateTime.UtcNow, UpdatedAt = DateTime.UtcNow };
            db.SmsPreferences.Add(p);
            await db.SaveChangesAsync();
        }
        return p;
    }
}

File: api/app/Http/Controllers/Api/SmsController.php
<?php
// app/Http/Controllers/Api/SmsController.php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use App\Models\SmsLog;
use App\Models\SmsPreference;
use App\Services\Sms\SmsSender;
use App\Support\Phone;
use App\Support\TenantContext;
use Illuminate\Http\Request;

class SmsController extends Controller
{
    public function balance(Request $request)
    {
        $tenant = $request->user()->tenant;

        return response()->json([
            'balance'   => $tenant->sms_balance,
            'sent_total' => SmsLog::count(),
        ]);
    }

    public function history(Request $request)
    {
        return SmsLog::query()->orderByDesc('id')
            ->paginate(min((int) $request->integer('per_page', 25), 100));
    }

    public function send(Request $request, SmsSender $sms)
    {
        $data = $request->validate([
            'phone'   => ['required', 'string'],
            'message' => ['required', 'string', 'max:600'],
        ]);

        $tenant = $request->user()->tenant;
        abort_if($tenant->sms_balance <= 0, 402, 'SMS bakiyeniz yetersiz.');

        $phone = Phone::normalize($data['phone']);
        $sms->send($phone, $data['message'], 'manual');

        $tenant->decrement('sms_balance');

        return response()->json(['message' => 'SMS gönderildi.', 'remaining' => $tenant->fresh()->sms_balance]);
    }

    public function getPreferences(Request $request)
    {
        $pref = SmsPreference::firstOrCreate(['tenant_id' => $request->user()->tenant_id]);

        return response()->json($pref);
    }

    public function updatePreferences(Request $request)
    {
        $data = $request->validate([
            'maintenance_reminder'  => ['nullable', 'boolean'],
            'maintenance_completed' => ['nullable', 'boolean'],
            'new_fault'             => ['nullable', 'boolean'],
            'fault_resolved'        => ['nullable', 'boolean'],
            'invoice_created'       => ['nullable', 'boolean'],
            'payment_received'      => ['nullable', 'boolean'],
            'tse_expiry'            => ['nullable', 'boolean'],
            'reminder_days_before'  => ['nullable', 'integer', 'min:0', 'max:30'],
        ]);

        $pref = SmsPreference::firstOrCreate(['tenant_id' => $request->user()->tenant_id]);
        $pref->update($data);

        return response()->json($pref->fresh());
    }
}
